dash/include/dash/halo/Halo.h

```diff
@@ -0,0 +1,1062 @@
+    for (dim_t i(0); i < NumDimensions; ++i)
+      max = std::max((int) max, (int) std::abs(this->_values[i]));
+    return max;
+  }
+}; // Stencil
+
+
+
+template <dim_t NumDimensions, std::size_t NumStencilPoints>
+class StencilSpec {
+
+private:
+  using SelfT    = StencilSpec<NumDimensions, NumStencilPoints>;
+  using StencilT = Stencil<NumDimensions>;
+  using SpecsT   = std::array<StencilT, NumStencilPoints>;
+
+public:
+  constexpr StencilSpec(const SpecsT& specs) : _specs(specs) {}
+
+  // TODO constexpr
+  StencilSpec(const SelfT& other) { _specs = other._specs; }
+
+  constexpr const SpecsT& stencilSpecs() const { return _specs; }
+
+  static constexpr std::size_t numStencilPoints() { return NumStencilPoints; }
+
+  constexpr const StencilT& operator[](std::size_t index) const { return _specs[index]; }
+
+private:
+  SpecsT _specs{};
+}; // StencilSpec
+
+
+
+/**
+ * Specifies the values global boundary Halos
+ */
+enum class Cycle : uint8_t {
+  /// No global boundary Halos
+  NONE,
+  /// Global boundary Halos with values from the opposite boundary
+  CYCLIC,
+  /// Global boundary Halos with predefined fixed values
+  FIXED
+};
+
+
+
+/**
+ * Cycle specification for every dimension
+ */
+template <dim_t NumDimensions>
+class CycleSpec : public Dimensional<Cycle, NumDimensions> {
+
+private:
+  using BaseT = Dimensional<Cycle, NumDimensions>;
+
+public:
+  // TODO constexpr
+  CycleSpec() {
+    for (dim_t i = 0; i < NumDimensions; ++i) {
+      this->_values[i] = Cycle::NONE;
+    }
+  }
+
+  template <typename... Values>
+  constexpr CycleSpec(Cycle value, Values... values) : BaseT::Dimensional(value, values...) {}
+}; // CycleSpec
+
+
+
+template <dim_t NumDimensions>
+class RegionCoords : public Dimensional<uint8_t, NumDimensions> {
+
+public:
+  using SelfT   = RegionCoords<NumDimensions>;
+  using coord_t = uint8_t;
+  using region_index_t = uint32_t;
+  using extent_t       = uint16_t;
+
+private:
+  using BaseT   = Dimensional<coord_t, NumDimensions>;
+  using CoordsT = std::array<coord_t, NumDimensions>;
+
+public:
+  RegionCoords() {
+    for (dim_t i = 0; i < NumDimensions; ++i) {
+      this->_values[i] = 1;
+    }
+    _index = index(this->_values);
+  }
+
+  template <typename... Values>
+  RegionCoords(coord_t value, Values... values) : BaseT::Dimensional(value, values...) {
+    _index = index(this->_values);
+  }
+
+  RegionCoords(region_index_t index) : _index(index) {
+    this->_values = coords(index);
+  }
+
+  constexpr region_index_t index() const { return _index; }
+
+  static region_index_t index(const CoordsT& coords) {
+    region_index_t index = coords[0];
+    for (auto i = 1; i < NumDimensions; ++i)
+      index = coords[i] + index * 3;
+
+    return index;
+  }
+
+  static CoordsT coords(const region_index_t index) {
+    CoordsT coords{};
+    region_index_t  index_tmp = static_cast<long>(index);
+    for (auto i = (NumDimensions - 1); i >= 1; --i) {
+      auto res  = std::div(index_tmp, 3);
```

Comment:
Where does the constant `3` come from?

Reply:
it's i fixed constant -> i have dimension^3 regions(including the center)
e.g. 1D 3 regions: north, center, south
       2D 9 regions: north west, north, north east, west, center, east,south west, south, south east
etc.

Reply:
Mhh, but in 1D 1^3 is 1 ;)

In any case, I would suggest using a `constexpr` variable to give the value `3` a meaningful name.

Reply:
meant 3^dimension. sorry

Reply:
constexpr THREE = 3 :D

Reply:
you are right. i will find a meaningful name